File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int ans = 0;
        int n = nums.length;
        int[] memo = new int[n+1];
        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, dfs(i, nums, memo));
        }

        return ans;
    }

    private int dfs(int n, int[] nums,int[] memo) {
        int ans = 0;
        if (memo[n] != 0) {
            return memo[n];
        } 
        for (int i = 0 ; i < n; i++) {
            if (nums[i] < nums[n]) {
                // 這邊因為是正序開始 所以沒有 i -1 
               ans = Math.max(ans, dfs(i, nums, memo));
            }
        }
        return memo[n] = ans + 1;
    }
}

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-2.py
from functools import cache
DIRECTIONS = [(1, 0), (-1, 0), (0, -1), (0, 1)]
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        
        @cache
        def dfs(i, j):
            if i < 0 or i >= m or j < 0 or j >= n:
                return 0
            ans = 1
            for dr, dc in DIRECTIONS:
                nr = i + dr
                nc = j + dc
                if 0 <= nr < m and 0 <= nc < n and matrix[i][j] < matrix[nr][nc]:
                    ans = max(ans, 1 + dfs(nr, nc))
            return ans

        ans = 0    
        for i in range(m):
            for j in range(n):
                    ans = max(ans, dfs(i, j))
        return ans            
